common/src/main/java/org/corfudb/common/util/FileWatcher.java

```diff
@@ -0,0 +1,103 @@
+package org.corfudb.common.util;
+
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.time.Duration;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+@Slf4j
+public class FileWatcher implements Closeable {
+
+    private final File file;
+
+    private final Runnable onChange;
+
+    private final ScheduledExecutorService watcher;
+
+    private volatile WatchService watchService;
+
+
+    public FileWatcher(String filePath, Runnable onChange, Duration pollPeriod){
+        this.file = Paths.get(filePath).toFile();
+        this.onChange = onChange;
+
+        this.watcher = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                        .setNameFormat("FileWatcher-")
+                        .build());
+
+        reloadNewWatchService();
+        watcher.scheduleAtFixedRate(
+                this::poll,
+                pollPeriod.toMillis(),
+                pollPeriod.toMillis(),
+                TimeUnit.MILLISECONDS);
+    }
+
+    private void poll() {
+        try {
+            WatchKey key = watchService.poll();
```

Comment:
Good point. I did that in my first version of the PR but I later realized that using a self managed scheduler to poll the watch service gives us better control. For example you can stop the scheduler explicitly instead of interrupting the blocking take() or poll(timeout).